Zadanie 1

a) Ściągnij plik iris.csv i umieść go w katalogu z plikami pythonowymi.
b) Użyj paczki pandas do wczytania tego pliku do Pythona i wyświetl bazę danych.
c) W obiekcie df mamy bazę danych typu dataframe. Żeby bezpośrednio wejść
do danych z tej bazy trzeba odwołać się do wartości (values).

In [31]:
import pandas as pd
import numpy as np

df = pd.read_csv("iris.csv")

print(df.values)

[[5.1 3.5 1.4 0.2 'Setosa']
 [4.9 3.0 1.4 0.2 'Setosa']
 [4.7 3.2 1.3 0.2 'Setosa']
 [4.6 3.1 1.5 0.2 'Setosa']
 [5.0 3.6 1.4 0.2 'Setosa']
 [5.4 3.9 1.7 0.4 'Setosa']
 [4.6 3.4 1.4 0.3 'Setosa']
 [5.0 3.4 1.5 0.2 'Setosa']
 [4.4 2.9 1.4 0.2 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [5.4 3.7 1.5 0.2 'Setosa']
 [4.8 3.4 1.6 0.2 'Setosa']
 [4.8 3.0 1.4 0.1 'Setosa']
 [4.3 3.0 1.1 0.1 'Setosa']
 [5.8 4.0 1.2 0.2 'Setosa']
 [5.7 4.4 1.5 0.4 'Setosa']
 [5.4 3.9 1.3 0.4 'Setosa']
 [5.1 3.5 1.4 0.3 'Setosa']
 [5.7 3.8 1.7 0.3 'Setosa']
 [5.1 3.8 1.5 0.3 'Setosa']
 [5.4 3.4 1.7 0.2 'Setosa']
 [5.1 3.7 1.5 0.4 'Setosa']
 [4.6 3.6 1.0 0.2 'Setosa']
 [5.1 3.3 1.7 0.5 'Setosa']
 [4.8 3.4 1.9 0.2 'Setosa']
 [5.0 3.0 1.6 0.2 'Setosa']
 [5.0 3.4 1.6 0.4 'Setosa']
 [5.2 3.5 1.5 0.2 'Setosa']
 [5.2 3.4 1.4 0.2 'Setosa']
 [4.7 3.2 1.6 0.2 'Setosa']
 [4.8 3.1 1.6 0.2 'Setosa']
 [5.4 3.4 1.5 0.4 'Setosa']
 [5.2 4.1 1.5 0.1 'Setosa']
 [5.5 4.2 1.4 0.2 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [5.0 3.2 1.2 0.2 'S

d) Spróbujmy wyświetlić różne fragmenty danych.

In [3]:
#wszystkie wiersze, kolumna nr 0
print(df.values[:, 0])
#wiersze od 5 do 10, wszystkie kolumny
print(df.values[5:11, :])
#dane w komórce [1,4]
print(df.values[1, 4])

[5.1 4.9 4.7 4.6 5.0 5.4 4.6 5.0 4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.0 5.0 5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.0
 5.5 4.9 4.4 5.1 5.0 4.5 4.4 5.0 5.1 4.8 5.1 4.6 5.3 5.0 7.0 6.4 6.9 5.5
 6.5 5.7 6.3 4.9 6.6 5.2 5.0 5.9 6.0 6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
 6.3 6.1 6.4 6.6 6.8 6.7 6.0 5.7 5.5 5.5 5.8 6.0 5.4 6.0 6.7 6.3 5.6 5.5
 5.5 6.1 5.8 5.0 5.6 5.7 5.7 6.2 5.1 5.7 6.3 5.8 7.1 6.3 6.5 7.6 4.9 7.3
 6.7 7.2 6.5 6.4 6.8 5.7 5.8 6.4 6.5 7.7 7.7 6.0 6.9 5.6 7.7 6.3 6.7 7.2
 6.2 6.1 6.4 7.2 7.4 7.9 6.4 6.3 6.1 7.7 6.3 6.4 6.0 6.9 6.7 6.9 5.8 6.8
 6.7 6.7 6.3 6.5 6.2 5.9]
[[5.4 3.9 1.7 0.4 'Setosa']
 [4.6 3.4 1.4 0.3 'Setosa']
 [5.0 3.4 1.5 0.2 'Setosa']
 [4.4 2.9 1.4 0.2 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [5.4 3.7 1.5 0.2 'Setosa']]
Setosa


e) Pora na podział zbioru na zbiór testowy i treningowy. Wykorzystamy do tego
funkcję train_test_split z paczki sklearn. Dodaj na górze programu odpowiedni

In [4]:
from sklearn.model_selection import train_test_split

Następnie dopisz na dole programu funckję dzielącą zbiór df.values na dwa
kawałki w proporcjach 70% i 30%, w losowy sposób.

In [5]:
#podzial na zbior testowy (30%) i treningowy (70%), ziarno losowosci = 13
(train_set, test_set) = train_test_split(df.values, train_size=0.7,
                                         random_state=274929)

f) Sprawdź jakie rekordy znalazły się w zbiorze testowym i ile ma on rekordów.

In [6]:
print(test_set)
print(test_set.shape[0])

[[6.1 2.8 4.7 1.2 'Versicolor']
 [6.9 3.1 4.9 1.5 'Versicolor']
 [5.0 3.4 1.6 0.4 'Setosa']
 [4.9 2.5 4.5 1.7 'Virginica']
 [5.7 3.8 1.7 0.3 'Setosa']
 [6.1 2.6 5.6 1.4 'Virginica']
 [5.6 2.9 3.6 1.3 'Versicolor']
 [6.4 3.1 5.5 1.8 'Virginica']
 [5.8 2.7 5.1 1.9 'Virginica']
 [4.7 3.2 1.3 0.2 'Setosa']
 [4.4 3.0 1.3 0.2 'Setosa']
 [5.2 2.7 3.9 1.4 'Versicolor']
 [5.1 3.4 1.5 0.2 'Setosa']
 [4.8 3.4 1.6 0.2 'Setosa']
 [6.3 2.5 4.9 1.5 'Versicolor']
 [5.8 2.7 4.1 1.0 'Versicolor']
 [7.7 3.0 6.1 2.3 'Virginica']
 [6.2 3.4 5.4 2.3 'Virginica']
 [6.7 3.0 5.0 1.7 'Versicolor']
 [6.7 3.3 5.7 2.1 'Virginica']
 [5.8 2.8 5.1 2.4 'Virginica']
 [6.3 3.4 5.6 2.4 'Virginica']
 [5.8 2.6 4.0 1.2 'Versicolor']
 [5.5 2.5 4.0 1.3 'Versicolor']
 [6.3 2.5 5.0 1.9 'Virginica']
 [4.7 3.2 1.6 0.2 'Setosa']
 [6.0 3.4 4.5 1.6 'Versicolor']
 [5.0 2.0 3.5 1.0 'Versicolor']
 [4.5 2.3 1.3 0.3 'Setosa']
 [7.2 3.6 6.1 2.5 'Virginica']
 [6.6 3.0 4.4 1.4 'Versicolor']
 [4.8 3.4 1.9 0.2 'Setosa']
 [4.4 2.9 1.4 0.2 'Seto

g) Bywa, że zbiór testowy i treningowy trzeba podzielić dodatkowo na inputy
(kolumny numeryczne) i klasy (odmiany irysów). Wówczas z 2 zbiorów danych
zrobią się nam cztery.

In [7]:
train_inputs = train_set[:, 0:4]
train_classes = train_set[:, 4]
test_inputs = test_set[:, 0:4]
test_classes = test_set[:, 4]

Zadanie 2

a) Rozpocznij od przygotowania zbioru treningowego i testowego, tak jak w
poprzednim zadaniu.

In [8]:
import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv("iris.csv")
(train_set, test_set) = train_test_split(df.values, train_size=0.7,
                                         random_state=274929)

b) Napiszemy w Pythonie funkcję, która na podstawie czterech numerycznych
parametrów irysa odgadnie jego gatunek wykorzystując do tego jedynie dwie
instrukcje if, elif oraz else (jednokrotnie).

In [9]:
def classify_iris(sl, sw, pl, pw):
    if sl > 4:
        return("Setosa")
    elif pl <= 5:
        return("Virginica")
    else:
        return("Versicolor")

c)

In [10]:
good_predictions = 0
len = test_set.shape[0]

for i in range(len):
    if classify_iris(test_inputs[i][0], test_inputs[i][1], test_inputs[i][2], test_inputs[i][3]) == test_set[i][4]:
        good_predictions += 1

print(good_predictions)
print(good_predictions/len*100, "%")

17
37.77777777777778 %


Wyniki:
17
37.77777777777778 %

d) Poprawiony klasyfikator.

In [11]:
def classify_iris2(sl, sw, pl, pw):
    if pw < 1:
        return("Setosa")
    elif pl > 4.8 and pw > 1.7:
        return("Virginica")
    else:
        return("Versicolor")

In [12]:
#sorted
print(train_set[train_set[:, 4].argsort()])

[[5.0 3.6 1.4 0.2 'Setosa']
 [5.2 4.1 1.5 0.1 'Setosa']
 [5.0 3.5 1.6 0.6 'Setosa']
 [4.8 3.0 1.4 0.3 'Setosa']
 [5.5 4.2 1.4 0.2 'Setosa']
 [5.1 3.8 1.9 0.4 'Setosa']
 [5.1 3.5 1.4 0.2 'Setosa']
 [5.1 3.8 1.6 0.2 'Setosa']
 [4.8 3.0 1.4 0.1 'Setosa']
 [4.6 3.2 1.4 0.2 'Setosa']
 [4.8 3.1 1.6 0.2 'Setosa']
 [5.4 3.4 1.7 0.2 'Setosa']
 [5.0 3.2 1.2 0.2 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [5.4 3.9 1.7 0.4 'Setosa']
 [5.2 3.4 1.4 0.2 'Setosa']
 [5.1 3.7 1.5 0.4 'Setosa']
 [5.7 4.4 1.5 0.4 'Setosa']
 [4.4 3.2 1.3 0.2 'Setosa']
 [5.4 3.4 1.5 0.4 'Setosa']
 [5.1 3.3 1.7 0.5 'Setosa']
 [5.8 4.0 1.2 0.2 'Setosa']
 [4.9 3.0 1.4 0.2 'Setosa']
 [5.0 3.3 1.4 0.2 'Setosa']
 [5.0 3.0 1.6 0.2 'Setosa']
 [4.3 3.0 1.1 0.1 'Setosa']
 [5.3 3.7 1.5 0.2 'Setosa']
 [5.2 3.5 1.5 0.2 'Setosa']
 [5.4 3.9 1.3 0.4 'Setosa']
 [5.1 3.8 1.5 0.3 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [4.9 3.1 1.5 0.1 'Setosa']
 [4.6 3.6 1.0 0.2 'Setosa']
 [6.0 2.7 5.1 1.6 'Versicolor']
 [6.7 3.1 4.4 1.4 'Versicolor']
 [6.0 2.2 4.

In [13]:
good_predictions2 = 0
len = test_set.shape[0]

for i in range(len):
    if classify_iris2(test_inputs[i][0], test_inputs[i][1], test_inputs[i][2], test_inputs[i][3]) == test_set[i][4]:
        good_predictions2 += 1

print(good_predictions2)
print(good_predictions2/len*100, "%")

42
93.33333333333333 %


Wyniki:
42
93.3%

zadanie 3

a) Podziel w losowy sposób bazę danych irysów na zbiór treningowy i zbiór
testowy w proporcjach 70%/30%. Wyświetl oba zbiory. Podziel te zbiory na
cztery części (inputy i class), jeśli jest taka potrzeba.

In [14]:
df = pd.read_csv("iris.csv")
(train_set, test_set) = train_test_split(df.values, train_size=0.7,
                                         random_state=274929)

train_inputs = train_set[:, 0:4]
train_classes = train_set[:, 4]
test_inputs = test_set[:, 0:4]
test_classes = test_set[:, 4]

b) Zainicjuj drzewo decyzyjne metodą DecisionTreeClassifier.

In [15]:
from sklearn.tree import DecisionTreeClassifier

dtc = DecisionTreeClassifier()

c) Wytrenuj drzewo decyzyjne na zbiorze treningowym, wykorzystując funkcję fit.

In [16]:
dtc.fit(train_inputs, train_classes)

DecisionTreeClassifier()

d) Wyświetl drzewo w formie tekstowej i/lub w formie graficznej. Jeśli masz
problemy z tym podpunktem – pomiń go.

In [17]:
from sklearn.tree import export_text

feature_names = ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

tree_to_print = export_text(dtc, feature_names=feature_names)
print(tree_to_print)

|--- petal length (cm) <= 2.45
|   |--- class: Setosa
|--- petal length (cm) >  2.45
|   |--- petal length (cm) <= 4.85
|   |   |--- petal width (cm) <= 1.70
|   |   |   |--- class: Versicolor
|   |   |--- petal width (cm) >  1.70
|   |   |   |--- sepal length (cm) <= 6.05
|   |   |   |   |--- class: Versicolor
|   |   |   |--- sepal length (cm) >  6.05
|   |   |   |   |--- class: Virginica
|   |--- petal length (cm) >  4.85
|   |   |--- petal width (cm) <= 1.70
|   |   |   |--- petal width (cm) <= 1.55
|   |   |   |   |--- class: Virginica
|   |   |   |--- petal width (cm) >  1.55
|   |   |   |   |--- sepal length (cm) <= 6.60
|   |   |   |   |   |--- class: Versicolor
|   |   |   |   |--- sepal length (cm) >  6.60
|   |   |   |   |   |--- class: Virginica
|   |   |--- petal width (cm) >  1.70
|   |   |   |--- class: Virginica



e) Dokonaj ewaluacji klasyfikatora: sprawdź jak drzewo poradzi sobie z rekordami
ze zbioru testowego. Wyświetl dokładność klasyfikatora, czyli procent
poprawnych odpowiedzi. Wykorzystaj do tego funkcję score lub predict.

In [18]:
test_score = dtc.score(test_inputs, test_classes)
print(test_score)
test_predict = dtc.predict(test_inputs)
print(test_predict)

0.8888888888888888
['Versicolor' 'Virginica' 'Setosa' 'Versicolor' 'Setosa' 'Virginica'
 'Versicolor' 'Virginica' 'Virginica' 'Setosa' 'Setosa' 'Versicolor'
 'Setosa' 'Setosa' 'Virginica' 'Versicolor' 'Virginica' 'Virginica'
 'Virginica' 'Virginica' 'Virginica' 'Virginica' 'Versicolor' 'Versicolor'
 'Virginica' 'Setosa' 'Versicolor' 'Versicolor' 'Setosa' 'Virginica'
 'Versicolor' 'Setosa' 'Setosa' 'Virginica' 'Versicolor' 'Setosa' 'Setosa'
 'Setosa' 'Setosa' 'Setosa' 'Versicolor' 'Setosa' 'Setosa' 'Versicolor'
 'Versicolor']


f) Wyświetl macierz błędów (ang. confusion matrix), która zestawia liczby
błędnych i poprawnych odpowiedzi, dla wszystkich klas.

In [19]:
from sklearn.metrics import confusion_matrix

confusion_matrix(test_classes, test_predict, labels=["Setosa", "Virginica", "Versicolor"])

array([[17,  0,  0],
       [ 0, 11,  2],
       [ 0,  3, 12]], dtype=int64)

Czy drzewo nauczone automatycznie jest lepsze niż Twoje drzewko stworzone ręcznie
w zadaniu 2?
Tak, Tutaj udało się osiągnąć 97.7%.
Ja osiągnąłem 93.3%.

Zadanie 4

Najpierw mój klasyfikator:

In [62]:
def classify_diabetes(pt, gc, bp, st, insulin, mi, pf, age):
    if age > 35 and gc > 123:
        return "tested_positive"
    else:
        return "tested_negative"

In [63]:
df2 = pd.read_csv("diabetes.csv")
(train_set2, test_set2) = train_test_split(df2.values, train_size=0.7,
                                         random_state=274929)

train_inputs2 = train_set2[:, 0:8]
train_classes2 = train_set2[:, 8]
test_inputs2 = test_set2[:, 0:8]
test_classes2 = test_set2[:, 8]

good_predictions = 0
len = test_set2.shape[0]

for i in range(len):
    if classify_diabetes(test_inputs2[i][0], test_inputs2[i][1], test_inputs2[i][2], test_inputs2[i][3], test_inputs2[i][4], test_inputs2[i][5], test_inputs2[i][6], test_inputs2[i][7]) == test_set2[i][8]:
        good_predictions += 1

print(good_predictions)
print(good_predictions/len*100, "%")

163
70.56277056277057 %


Wyniki:
163
70.56%

Wypisanie posortowanych danych dla ułatwienia znajdywania zależności gołym okiem.

In [32]:
# jest to potrzebne, by wyświetlić wszystkie rzędy
with np.printoptions(threshold=np.inf):
    print(train_set2[train_set2[:, 8].argsort()])

[[1 116 70 28 0 27.4 0.204 21 'tested_negative']
 [4 118 70 0 0 44.5 0.904 26 'tested_negative']
 [6 103 72 32 190 37.7 0.324 55 'tested_negative']
 [1 87 68 34 77 37.6 0.401 24 'tested_negative']
 [1 81 74 41 57 46.3 1.096 32 'tested_negative']
 [1 79 75 30 0 32.0 0.396 22 'tested_negative']
 [2 122 52 43 158 36.2 0.816 28 'tested_negative']
 [0 137 70 38 0 33.2 0.17 22 'tested_negative']
 [4 129 86 20 270 35.1 0.231 23 'tested_negative']
 [3 116 74 15 105 26.3 0.107 24 'tested_negative']
 [4 112 78 40 0 39.4 0.236 38 'tested_negative']
 [0 108 68 20 0 27.3 0.787 32 'tested_negative']
 [3 106 72 0 0 25.8 0.207 27 'tested_negative']
 [3 148 66 25 0 32.5 0.256 22 'tested_negative']
 [2 175 88 0 0 22.9 0.326 22 'tested_negative']
 [3 103 72 30 152 27.6 0.73 27 'tested_negative']
 [4 97 60 23 0 28.2 0.443 22 'tested_negative']
 [8 126 88 36 108 38.5 0.349 49 'tested_negative']
 [7 94 64 25 79 33.3 0.738 41 'tested_negative']
 [4 129 60 12 231 27.5 0.527 31 'tested_negative']
 [1 144 82 40

a) Podziel w losowy sposób bazę danych irysów na zbiór treningowy i zbiór
testowy w proporcjach 70%/30%. Wyświetl oba zbiory. Podziel te zbiory na
cztery części (inputy i class), jeśli jest taka potrzeba.

In [23]:
df2 = pd.read_csv("diabetes.csv")
(train_set2, test_set2) = train_test_split(df2.values, train_size=0.7,
                                         random_state=274929)


train_inputs2 = train_set2[:, 0:8]
train_classes2 = train_set2[:, 8]
test_inputs2 = test_set2[:, 0:8]
test_classes2 = test_set2[:, 8]

b) Zainicjuj drzewo decyzyjne metodą DecisionTreeClassifier

In [24]:
dtc2 = DecisionTreeClassifier()

c) Wytrenuj drzewo decyzyjne na zbiorze treningowym, wykorzystując funkcję fit.

In [25]:
dtc2.fit(train_inputs2, train_classes2)

DecisionTreeClassifier()

d) Wyświetl drzewo w formie tekstowej i/lub w formie graficznej. Jeśli masz
problemy z tym podpunktem – pomiń go.

In [26]:
from sklearn.tree import export_text

feature_names2 = ['spregnant-times', 'glucose-concentr', 'blood-pressure', 'skin-thickness', 'insulin', 'mass-index', 'pedigree-func', 'age']

tree_to_print2 = export_text(dtc2, feature_names=feature_names2)
print(tree_to_print2)

|--- glucose-concentr <= 129.50
|   |--- age <= 28.50
|   |   |--- mass-index <= 30.95
|   |   |   |--- spregnant-times <= 7.50
|   |   |   |   |--- glucose-concentr <= 128.50
|   |   |   |   |   |--- pedigree-func <= 0.67
|   |   |   |   |   |   |--- class: tested_negative
|   |   |   |   |   |--- pedigree-func >  0.67
|   |   |   |   |   |   |--- pedigree-func <= 0.69
|   |   |   |   |   |   |   |--- class: tested_positive
|   |   |   |   |   |   |--- pedigree-func >  0.69
|   |   |   |   |   |   |   |--- class: tested_negative
|   |   |   |   |--- glucose-concentr >  128.50
|   |   |   |   |   |--- skin-thickness <= 14.50
|   |   |   |   |   |   |--- class: tested_negative
|   |   |   |   |   |--- skin-thickness >  14.50
|   |   |   |   |   |   |--- class: tested_positive
|   |   |   |--- spregnant-times >  7.50
|   |   |   |   |--- class: tested_positive
|   |   |--- mass-index >  30.95
|   |   |   |--- blood-pressure <= 51.00
|   |   |   |   |--- insulin <= 75.50
|   |   |   |   |

e) Dokonaj ewaluacji klasyfikatora: sprawdź jak drzewo poradzi sobie z rekordami ze zbioru testowego. Wyświetl dokładność klasyfikatora, czyli procent poprawnych odpowiedzi. Wykorzystaj do tego funkcję score lub predict.

In [27]:
test_score2 = dtc2.score(test_inputs2, test_classes2)
print(test_score2)
test_predict2 = dtc2.predict(test_inputs2)
print(test_predict2)

0.7316017316017316
['tested_negative' 'tested_negative' 'tested_negative' 'tested_negative'
 'tested_negative' 'tested_positive' 'tested_negative' 'tested_negative'
 'tested_negative' 'tested_negative' 'tested_negative' 'tested_negative'
 'tested_negative' 'tested_positive' 'tested_negative' 'tested_negative'
 'tested_negative' 'tested_negative' 'tested_positive' 'tested_negative'
 'tested_negative' 'tested_negative' 'tested_positive' 'tested_negative'
 'tested_positive' 'tested_negative' 'tested_negative' 'tested_negative'
 'tested_negative' 'tested_positive' 'tested_positive' 'tested_positive'
 'tested_negative' 'tested_negative' 'tested_negative' 'tested_positive'
 'tested_positive' 'tested_negative' 'tested_positive' 'tested_negative'
 'tested_negative' 'tested_positive' 'tested_negative' 'tested_positive'
 'tested_negative' 'tested_positive' 'tested_negative' 'tested_positive'
 'tested_negative' 'tested_negative' 'tested_negative' 'tested_negative'
 'tested_negative' 'tested_negat

Wynik:
73.1%

f) Wyświetl macierz błędów (ang. confusion matrix), która zestawia liczby
błędnych i poprawnych odpowiedzi, dla wszystkich klas.

In [28]:
confusion_matrix(test_classes2, test_predict2, labels=['tested_positive', 'tested_negative'])

array([[ 46,  32],
       [ 30, 123]], dtype=int64)

Czy drzewo nauczone automatycznie jest lepsze niż Twoje drzewko stworzone ręcznie
w zadaniu 2? Ponownie Tak, choć tylko trochę.
Maszyna -> 73.1%
Moja funkcja -> 70.56%